File: lib/extraction/pdf-text-layer.ts
import { loadPdfjs, STANDARD_FONT_DATA_URL } from "@/lib/extraction/pdfjs-runtime";

export type TextLayerLine = {
  page: number; // 1-based
  text: string;
  box: [number, number, number, number]; // [x0,y0,x1,y1], normalized 0-1, top-left origin
};

/**
 * Reads the real text layer out of a PDF — page, string, and position for
 * every text-showing operation. This is the deterministic ground truth that
 * align-evidence.ts checks extracted values against. The model never gets to
 * report its own coordinates (CLAUDE.md section 10).
 */
export async function extractTextLayer(pdfBytes: Uint8Array): Promise<TextLayerLine[]> {
  const pdfjsLib = await loadPdfjs();
  // pdfjs's `data` option takes ownership of the buffer it's given and
  // detaches it (byteLength drops to 0 on every reference to the same
  // underlying ArrayBuffer, caller included) — pass a copy, never
  // `pdfBytes` itself, so callers can safely reuse their own buffer
  // afterward (the upload sandbox's validateUpload does exactly that, and
  // this was silently unsafe for every existing caller too, just never hit
  // because none of them happened to reuse the buffer after this call).
  const loadingTask = pdfjsLib.getDocument({ data: pdfBytes.slice(), standardFontDataUrl: STANDARD_FONT_DATA_URL });
  const pdf = await loadingTask.promise;

  const lines: TextLayerLine[] = [];

  for (let pageNum = 1; pageNum <= pdf.numPages; pageNum++) {
    const page = await pdf.getPage(pageNum);
    const viewport = page.getViewport({ scale: 1 });
    const content = await page.getTextContent();

    for (const item of content.items) {
      if (!("str" in item) || !item.str.trim()) continue;
      const [, , , , e, f] = item.transform as number[];
      const width = item.width ?? 0;
      const height = item.height ?? 10;

      const pdfTop = f + height;
      const pdfBottom = f;
      const x0 = clamp01(e / viewport.width);
      const x1 = clamp01((e + width) / viewport.width);
      const y0 = clamp01((viewport.height - pdfTop) / viewport.height);
      const y1 = clamp01((viewport.height - pdfBottom) / viewport.height);

      lines.push({ page: pageNum, text: item.str, box: [x0, y0, x1, y1] });
    }
  }

  return lines;
}

function clamp01(n: number): number {
  return Math.min(1, Math.max(0, n));
}

File: ledgerguard/extraction/pdf_text_layer.py
"""Reads the real text layer out of a PDF — page, string, and position for
every line-level text-showing operation. This is the deterministic ground
truth that align_evidence.py checks extracted values against. The model
never gets to report its own coordinates.

Reads the text layer with pdfplumber. `extract_text_lines()` clusters
characters into lines by layout, which matches the document's own
text-showing operations because pdf_generate.py draws exactly one operation
per document line — and pdfplumber's box coordinates are top-left-origin, so
dividing by page width/height yields normalized coordinates directly, with
no y-axis flip needed.
"""

import io
from dataclasses import dataclass

import pdfplumber


@dataclass
class TextLayerLine:
    page: int  # 1-based
    text: str
    box: tuple[float, float, float, float]  # (x0, y0, x1, y1), normalized 0-1, top-left origin


def _clamp01(n: float) -> float:
    return min(1.0, max(0.0, n))


def extract_text_layer(pdf_bytes: bytes) -> list[TextLayerLine]:
    lines: list[TextLayerLine] = []
    with pdfplumber.open(io.BytesIO(pdf_bytes)) as pdf:
        for page_num, page in enumerate(pdf.pages, start=1):
            width, height = page.width, page.height
            if width <= 0 or height <= 0:
                continue
            for raw_line in page.extract_text_lines(strip=True, return_chars=False):
                text = raw_line.get("text", "")
                if not text.strip():
                    continue
                x0 = _clamp01(raw_line["x0"] / width)
                x1 = _clamp01(raw_line["x1"] / width)
                y0 = _clamp01(raw_line["top"] / height)
                y1 = _clamp01(raw_line["bottom"] / height)
                lines.append(TextLayerLine(page=page_num, text=text, box=(x0, y0, x1, y1)))
    return lines


def count_text_items_per_page(pdf_bytes: bytes) -> list[int]:
    """Used by the upload sandbox's text-density check — a low count means a
    scanned image with no real text layer, which this app rejects rather
    than OCRing."""
    counts: list[int] = []
    with pdfplumber.open(io.BytesIO(pdf_bytes)) as pdf:
        for page in pdf.pages:
            counts.append(len(page.extract_words()))
    return counts
